feat(docs): add specifics about using our mock fixture (#603)
stravalib/stravalib: docs/contributing/development-guide.md

## Setup a local development environment

> We suggest you create a virtual environment on your computer to work on

Author: here i added instructions for creating a venv and used tabs for conda vs venv

## Python support
We loosely follow the [Numpy guidelines defined in NEP 29](https://numpy.org/neps/nep-0029-deprecation_policy.html)
for Python version support. However, in some cases, we may decide to
support older versions of Python, following community demand.

## Code style, linting & typing

> We use several tools to maintain consistent code formatting and adhere to the [Python Enhancement Protocol (PEP) 8](https://peps.python.org/pep-0008/) standards, which outline best practices for Python code readability and structure. Below are the primary tools configured for this project:

Author: Here i updated this section with our current lint and code format config.

> ## About the stravalib test suite

Author: This is where the significant updates are!

To run the test suite across all python versions that we support use:
To run the test suite across all Python versions that we support use:

```
nox -s tests
```

`nox -s tests` does a few things:

1. It create a temporary directory called `tmp-test-dir-stravalib` in which your tests are run. We create this test directory to ensure that tests are being run against the installed version of stravalib (with the most recent local development changes as installed) rather than the flat files located in the GitHub repository.
1. It creates a temporary directory called `tmp-test-dir-stravalib` in which your tests are run. We create this test directory to ensure that tests are being run against the installed version of stravalib (with the most recent local development changes as installed) rather than the flat files located in the GitHub repository.
2. It runs the tests and provides output (see below)
3. Finally it removes the temporary directory

To run tests for a specific python version use nox -s tests-python-version-here.

For example, below we run tests for only Python 3.10.

```bash
nox -s tests-3.10
```

> ### Functional end-to-end test suite

Author: We dropped the functional test suite a while back because it required a token and app, etc. so I removed this section.

```{note}
The actual code coverage report is
uploaded on the GitHub action run on `ubuntu` and `Python 3.11`. When that step in the
actions completes, the report will be processed and returned to the pull request.
```

## Tests & the stravalib mock fixture

To run integration tests that ensure stravalib is interacting with API data correctly, Stravalib uses a mock object access through a
`pytest` fixture `stravalib.tests.integration.strava_api_stub.StravaAPIMock`
that is based on `responses.RequestsMock`.

This fixture adds a pass-through mock that prevents requests from being made to the Strava API.
Instead, it creates responses using the endpoint provided and the `swagger.json` file that is found both online and within the `stravalib/src/stravalib/tests/resources/` directory that are based on examples from the published Strava API
documentation.

:::{tip}
Example usages of this fixture can be found in the
{py:mod}`stravalib.tests.integration.test_client` module.
:::

:::{mermaid}

flowchart TD
    B["conftest.py"] -- Defines fixture --> C["mock_strava_api fixture"]

>     C -- Calls --> D["StravaAPIMock from strava_api_stub.py module"]

Reviewer: ```suggestion
    C -- Creates instance of --> D["StravaAPIMock from strava_api_stub.py module"]
```

>     D -- Inherits from --> E["requests_mock"]

Reviewer: ```suggestion
    D -- Inherits from --> E["responses.RequestsMock"]
```

>     D -- Adds API Call passthru and returns data using: --> G["swagger.json <br>(local or online)"]

Reviewer: ```suggestion
    D -- Returns fake response data using: --> G["swagger.json <br>(local or online)"]
```

>     D -- Calls --> H["_api_method_adapter"]

Reviewer: I don't think (new) contributors need to know that level of detail. Maybe we can remove this single line?

Author: perfect. i simplified the diagram. i'll merge once CI is happy! thank you for the feedback!

If the object can find an endpoint match, it then returns the example JSON response
(or the updated response if you use the update parameter) to use in the test.

:::{tip}
The `swagger.json` file is an API specification document describing the
available endpoints in the Strava API, including methods, parameters, and
expected responses for each endpoint. It defines the API structure in JSON
format and includes example responses for testing. This file is used in
`stravalib`'s tests to mock API interactions and validate the expected
structure and content of responses.

The mock API object checks if `swagger.json` is accessible online; if not,
it uses a local version located in the `tests/resources` directory within
`stravalib`.
:::

### Mock fixture features

To call the mock fixture in a test, you

1. Create a new test and add the `mock_strava_api` fixture as an input to the test function.

The test below will try to access the `/athlete/activities`
Strava endpoint which returns an athlete's activities.
Here, the fixture will bypass trying to access the real online API. And instead, will find the `/athlete/activities`
endpoint in the Strava online or local `swagger.json` file.

> When you call {py:func}`stravalib.client.Client.get_activities()`, the endpoint will return the sample data provided in the `swagger.json` file.

Reviewer: ```suggestion
When you call {py:func}`stravalib.client.Client.get_activities()`, the mocked endpoint will return the sample data provided in the `swagger.json` file.
```

    assert len(activity_list) == 4
```

The mock fixture object provides parameters that allow you to modify a test.

Sometimes you may want to update the default example return data in the swagger.json file. This might happen if you want to intentionally "break" a test to ensure that the client call responds appropriately.

To modify the returned sample data use the `response_update`
parameter. Below you update the response id key to be another value.

```python
def test_example_test(mock_strava_api, client):
    """An example test"""
    mock_strava_api.get(
        "/athlete/activities",
        response_update={"id": 12345},
    )
    activity_list = list(client.get_activities())
```

You can also specify the number of results that you'd like to see in the mock output using the `n_results` parameters.

```python
def test_example_test(mock_strava_api, client):
    """An example test"""
    mock_strava_api.get(
        "/athlete/activities",
        response_update={"id": 12345},
        n_results=4,
    )
    activity_list = list(client.get_activities())
```

:::{tip}
Stravalib uses lazily loaded entities when returning results from

> endpoint such as activities that may include multiple response objects in the return. As such, a mocked call to {py:func}`stravalib.client.Client.get_activities` will not actually initiative a get response

Reviewer: ```suggestion
endpoint such as activities that may include multiple response objects in the return. As such, a mocked call to {py:func}`stravalib.client.Client.get_activities` will not actually initiate a get response
```
